factor/grace/settings.php

```diff
@@ -37,6 +37,18 @@
     new lang_string('settings:forcesetup', 'factor_grace'),
     new lang_string('settings:forcesetup_help', 'factor_grace'), 0));
 
-$name = new lang_string('settings:graceperiod', 'factor_grace');
-$description = new lang_string('settings:graceperiod_help', 'factor_grace');
-$settings->add(new admin_setting_configduration('factor_grace/graceperiod', $name, $description, '604800'));
+$settings->add(new admin_setting_configduration('factor_grace/graceperiod',
+    new lang_string('settings:graceperiod', 'factor_grace'),
+    new lang_string('settings:graceperiod_help', 'factor_grace'), '604800'));
+
+// Get factors that gracemode cares about.
+$activefactors = \tool_mfa\plugininfo\factor::get_enabled_factors();
+$active = [];
+foreach ($activefactors as $factor) {
+    if ($factor->has_setup() || $factor->has_input()) {
```

Comment:
why do we do it here?

Reply:
Gracemode only cares about factors that a user can input or affect. Filtering here makes the list a lot shorter and less cluttered.